crypto/params.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+
+#define SET_RETURN_SIZE(p, sz) \
+    if (p->return_size != NULL) \
+        *p->return_size = (sz)
+
+static const OSSL_PARAM *param_locate(const OSSL_PARAM *p, const char *key)
+{
+    if (p != NULL && key != NULL)
+        for (; p->key != NULL; p++)
+            if (strcmp(key, p->key) == 0)
+                return p;
+    return NULL;
+}
+
+#define PARAM_INT_GET(to, from) \
+    case sizeof(from): \
+        if (sizeof(to) >= sizeof(from)) { \
+            *val = (to)*(const from *)p->buffer; \
+            return 1; \
+        } \
+        break
+
+#define PARAM_INT_SET(to, from) \
+    case sizeof(to): \
+        SET_RETURN_SIZE(p, sizeof(to)); \
+        if (sizeof(to) >= sizeof(from)) { \
+            *(to *)p->buffer = (to)val; \
+            return 1; \
+        } \
+        break
+
+#define PARAM_INT(name, type, pre) \
+    int OSSL_PARAM_get_##name(const OSSL_PARAM *p, const char *key, type *val) \
+    { \
+        if (val == NULL \
+            || (p = param_locate(p, key)) == NULL \
+            || (p->data_type != OSSL_PARAM_INTEGER \
+                && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER)) \
+            return 0; \
+        switch (p->buffer_size) { \
+        PARAM_INT_GET(type, pre ## 32_t); \
+        PARAM_INT_GET(type, pre ## 64_t); \
+        } \
+        return 0; \
+    } \
+    int OSSL_PARAM_set_##name(const OSSL_PARAM *p, const char *key, type val) \
+    { \
+        if ((p = param_locate(p, key)) == NULL) \
+            return 0; \
+        if (p->data_type == OSSL_PARAM_INTEGER \
+            || p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) \
+            switch (p->buffer_size) { \
+            PARAM_INT_SET(pre ## 32_t, type); \
+            PARAM_INT_SET(pre ## 64_t, type); \
+            } \
+        SET_RETURN_SIZE(p, 0); \
```

Comment:
... then remove this line.

Reply:
(at least to my eye, this reset any other value set through `PARAM_INT_SET`)

Reply:
There is the same issue in set_double below.  I'll fix that as well.